### PR TITLE
ENH: Only apply first group once in fast GroupBy.apply

### LINKED ISSUE #2936: ENH: option to force slow code path (don't call apply function 1 too many times) in GroupBy.apply
```
>>> def applym(df):
...     print df.irow(0)['a']
...     return DataFrame({'a':[1],'b':[1]})

>>> df = DataFrame({'a':[1,1,1,2,3],'b':['a','a','a','b','c']})

>>> df.groupby('b').apply(applym)
1
1
2
3
     a  b
b        
a 0  1  1
b 0  1  1
c 0  1  1
>>>
```

applym is called twice on the first group.

### LINKED ISSUE #10519: GroupBy.apply calling function twice
```python
def foo(x):
    print "hello"
_ = pd.DataFrame({'a':[1]}).groupby(["a"]).apply(foo)
```
```
hello
hello
```

### LINKED ISSUE #12155: BUG: GroupBy.apply iterates over first group twice
When using the ```GroupBy.apply()``` method with a custom function handle, the first group is iterated over twice. Explicitly looping over each group and applying the function directly does not yield the error.

Error reproduction
```python
>>> import pandas as pd
>>> df = pd.DataFrame({'one':[1,2,3,1,2,3],'two':[4,5,6,7,8,9]})
>>> df
   one  two
0    1    4
1    2    5
2    3    6
3    1    7
4    2    8
5    3    9

>>> gp = df.groupby('one')
>>> def foo(group):
...   print(group.iloc[[0]].index[0])
...
>>> gp.apply(foo)
0
0
1
2
Empty DataFrame
Columns: []
Index: []

>>> len(gp)
3
```

The example prints the row index of the first group member. The '0' is printed twice.

Versions:
```python
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 58 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.17.1
nose: None
pip: 8.0.1
setuptools: 19.4
Cython: None
numpy: 1.10.1
scipy: None
statsmodels: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
Jinja2: None
```

### LINKED ISSUE #20084: ENH: Add argument to GroupBy.apply to let user pre-select "slow path" & not run function twice on 1st group
I have a large data set that includes a "category" column. I'm working to use the data to build several models, one for each category in the data (5 categories total). I'd like to use `groupby` to do this, and I wrote a custom function that can be used in `GroupBy.apply`. It works fine, although the function is quite slow when it's just building one model, let alone 5. Unfortunately, because of the way `GroupBy.apply` works, it's actually building 6 models instead of 5. As a result, my code using `GroupBy.apply` is 20% slower than it would be if I'd just written a for loop.

As [the documentation for GroupBy.apply](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.core.groupby.GroupBy.apply.html) states:

> In the current implementation `apply` calls func twice on the first group to decide whether it can take a fast or slow code path. This can lead to unexpected behavior if func has side-effects, as they will take effect twice for the first group.

Since I know that my function is going to be slow anyway, I'm wondering if it's possible to add an argument to `apply` that will let me opt to take the "slow code path" from the start, and skip the test-run of the function on the first group? When my custom function takes 20 minutes each time it runs anyway, being able to cut out the extra iteration would be quite nice, and I can't imagine that any time lost by taking the slow path over the fast path would make a difference. (Or am I completely wrong about that?)

Sure, I could just write a `for` loop and everything would work and be fine, but I really like the "neatness" of using `GroupBy`, if that makes sense. Additionally, it would be nice if I could use `GroupBy.apply` in a similar way to how [the `do` function in `dplyr`](http://dplyr.tidyverse.org/reference/do.html) works after grouping.

##### Current functionality

```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'A': list('aaabbbcccc'), 
                           'B': [3,4,3,6,5,2,1,9,5,4], 
                           'C': [4,0,2,2,2,7,8,6,2,8]})

In [3]: def print_name_and_describe(g):
   ...:     print(g.name)
   ...:     return g.describe()
   ...: 

In [4]: df.groupby('A').apply(print_name_and_describe)
a
a
b
c
Out[4]: 
                B         C
A                          
a count  3.000000  3.000000
  mean   3.333333  2.000000
  std    0.577350  2.000000
  min    3.000000  0.000000
  25%    3.000000  1.000000
  50%    3.000000  2.000000
  75%    3.500000  3.000000
  max    4.000000  4.000000
b count  3.000000  3.000000
  mean   4.333333  3.666667
  std    2.081666  2.886751
  min    2.000000  2.000000
  25%    3.500000  2.000000
  50%    5.000000  2.000000
  75%    5.500000  4.500000
  max    6.000000  7.000000
c count  4.000000  4.000000
  mean   4.750000  6.000000
  std    3.304038  2.828427
  min    1.000000  2.000000
  25%    3.250000  5.000000
  50%    4.500000  7.000000
  75%    6.000000  8.000000
  max    9.000000  8.000000
```

##### Suggested functionality

```python
In [4]: df.groupby('A').apply(print_name_and_describe, use_slow_path=True)
a
b
c
Out[4]: 
                B         C
A                          
a count  3.000000  3.000000
  mean   3.333333  2.000000
  std    0.577350  2.000000
  min    3.000000  0.000000
  25%    3.000000  1.000000
  50%    3.000000  2.000000
  75%    3.500000  3.000000
  max    4.000000  4.000000
b count  3.000000  3.000000
  mean   4.333333  3.666667
  std    2.081666  2.886751
  min    2.000000  2.000000
  25%    3.500000  2.000000
  50%    5.000000  2.000000
  75%    5.500000  4.500000
  max    6.000000  7.000000
c count  4.000000  4.000000
  mean   4.750000  6.000000
  std    3.304038  2.828427
  min    1.000000  2.000000
  25%    3.250000  5.000000
  50%    4.500000  7.000000
  75%    6.000000  8.000000
  max    9.000000  8.000000
```

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.13.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.22.0
pytest: None
pip: 9.0.1
setuptools: 36.5.0
Cython: None
numpy: 1.14.1
scipy: None
pyarrow: None
xarray: None
IPython: 5.5.0
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2018.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: 1.0b10
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### LINKED ISSUE #21417: Additional calls to function when using GroupBy.apply with multiple index
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
df = pd.DataFrame([[1, 2, 3], [2, 2, 3]], columns=['a', 'b', 'c'])
def t(df):
    print(df)
    print('-' * 10)
    return None

g = df.groupby(by=['a', 'b'])
g.apply(t)
print(g.groups)
g.apply(lambda x:print(x.index))
```
#### Problem description
The function is invoked by 'apply' three times while there are only two groups.

#### Expected Output
The function is only invoked once by 'apply' for each group.
#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-119-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_HK.UTF-8
LOCALE: en_HK.UTF-8

pandas: 0.21.0
pytest: 3.2.1
pip: 9.0.1
setuptools: 36.4.0
Cython: None
numpy: 1.14.3
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: 2.6.5
feather: None
matplotlib: None
openpyxl: 2.4.9
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999999999
sqlalchemy: 1.1.15
pymysql: None
psycopg2: 2.7.3.1 (dt dec pq3 ext lo64)
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by fjetter**
The issue of applying a function to the first row twice has been reported quite a few times (issues are usually referenced to #2936) and is also a documented shortcoming/implementation detail (see Notes in [docs](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.apply.html))

The argumentation is usually that this needs to be done to determine whether or not we can take a fast path for the calculation. Even if we're using the fast path for the calculation, however, the first row is still evaluated twice. I believe with this refactoring it would be possible to only eval the first row once iff the fast path is taken.
For functions with side effects or very expensive calculations in general this may be a big deal as discussed in various other issues.

I'm wondering if this small change may already help out folks. Truth be told, I'm not 100% certain about what's actually happening in this loop regarding the slider and item cache. For the use case where I encountered this, it seems to do the trick, though.

closes #2936
closes #2656
closes #7739
closes #10519
closes #12155
closes #20084
closes #21417

- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
I am not sure you actually need to change, this rather you can modify code in pandas/core/groupby/generic.py and pandas/core/groupby/ops.py, search for ``fast_path``. 

but much more important is to actually have some tests for this.

**Comment by fjetter** [comment]
I don't believe that changing the python code is actually sufficient since the Cython code I am modifying evaluates the first row twice, even if everything works as expected. My change here is essentially only using the calculation for the first row/test if the test is successful instead of recalculating it again.

I believe you are referring to https://github.com/pandas-dev/pandas/blob/453fa85a8b88ca22c7b878a3fcf97e068f11b6c4/pandas/core/groupby/ops.py#L174
where `fast_apply` is called and if successful, the result (calculated in Cython) is immediately returned. If an `InvalidApply` exception is caught, a slow, python-only path is taken. I'm not trying to optimize the slow/python path but I am arguing that even the `fast_apply` path will evaluate the first row twice although this isn't necessary.

Apart from implicit tests (this code section should be called at many places) there is one explicit test to ensure that the fast path can be taken and the flags for `mutated` are returned as expected, see https://github.com/pandas-dev/pandas/blob/453fa85a8b88ca22c7b878a3fcf97e068f11b6c4/pandas/tests/groupby/test_apply.py#L81
If there are other edge cases you can think of I can add more tests.

**Review by jreback — CHANGES_REQUESTED** [review]
@fjetter this is not going to be acceptable w/o a test that fails before and passes now, meaning you have to show an explicit test that does this.

File: pandas/tests/groupby/test_groupby.py [hunk 1]
```diff
@@ -1397,7 +1397,7 @@ def f(group):
     df.groupby('a', sort=False, group_keys=False).apply(f)
     # we expect 2 zeros because we call ``f`` once to see if a faster route
     # can be used.
-    expected_names = [0, 0, 1, 2]
+    expected_names = [0, 1, 2]
```
Comment (fjetter) [inline pandas/tests/groupby/test_groupby.py]: @jreback This is the only test I am modifying in this PR. Essentially it tracks how often the function is applied to a given row/group by appending the groups name to a list. 
Previous behavior: Apply it twice to the first row. The first apply is done to check for `fast_apply` compatibility. The second apply is due to looping again from 0 to 2. 
In my change I propose to perform the initial `fast_apply` compatibility check _within_ the loop to save one iteration. 

Is this what you had in mind or what scenario would you like me to test?

**Comment by jreback** [comment]
@fjetter I would like to collect all tests from any of the references issues. After you add them see if this closes them; list the issues in the top of the PR.

File: pandas/tests/frame/test_apply.py [hunk 2]
```diff
@@ -568,6 +568,37 @@ def test_apply_dup_names_multi_agg(self):
 
         tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.parametrize("axis, expected", [
+        (0, ['a', 'b']),
+        (1, [0, 1, 2, 3, 4, 5]),
+    ])
+    def test_apply_first_row_once(self, axis, expected):
+        df = pd.DataFrame({'a': [0, 0, 1, 1, 2, 2], 'b': np.arange(6)})
```
Comment (jreback) [inline pandas/tests/frame/test_apply.py]: can you add the issue number as a comment

File: pandas/tests/frame/test_apply.py [hunk 3]
```diff
@@ -568,6 +568,37 @@ def test_apply_dup_names_multi_agg(self):
...
+
+        rows = []
+
+        def f_fast(row):
+            rows.append(row.name)
+            return 0
+        df.apply(f_fast, axis=axis)
+        # gh-2936
```
Comment (jreback) [inline pandas/tests/frame/test_apply.py]: ahh you have here, well move to the top

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a comment with the issue number on the tests (if the test applies to multiple issue, then list both of them).

Add a whatsnew entry in 0.25.0. I would make this a sub-section in api-breaking changes, showing a mini-example and listing all of the closed issues.

**Comment by jreback** [comment]
also we have a warning in the groupby.rst about this, can remove it.

**Comment by fjetter** [comment]
I added a whatsnew entry and added the issue references to the tests. I kept the warning in `groupby.rst` since the double func evaluation may still occur if we need to fall back to the slow path.

**Comment by jreback** [comment]
> I added a whatsnew entry and added the issue references to the tests. I kept the warning in groupby.rst since the double func evaluation may still occur if we need to fall back to the slow path.

that needs to be addressed here as well - i suspect this is actually the more common problem

**Comment by fjetter** [comment]
I changed the implementation such that also the slow path evaluates the function only once. I refactored `apply_frame_axis0` to signal the sucess/validity of the fast apply via a status code instead of an exception. If the status code is 1, we fall back to the slow path calculation but will use the already calculated result for the first group.

File: pandas/tests/groupby/test_groupby.py [hunk 4]
```diff
@@ -1420,20 +1420,30 @@ def foo(x):
 
 def test_group_name_available_in_inference_pass():
     # gh-15062
+    # GH24748 ,GH2936, GH2656, GH7739, GH10519, GH12155, GH20084, GH21417
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: don’t add all the issues, just the ones that apply to each test; then it should be easy to see that we have all of the issues covered

File: doc/source/whatsnew/v0.25.0.rst [hunk 5]
```diff
@@ -26,6 +26,51 @@ Other Enhancements
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+GroupBy.apply on ``DataFrame`` evaluates first group only once
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+(:issue:`2936`, :issue:`2656`, :issue:`7739`, :issue:`10519`, :issue:`12155`,
+:issue:`20084`, :issue:`21417`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: issues go at the end o the text

File: doc/source/whatsnew/v0.25.0.rst [hunk 6]
```diff
@@ -26,6 +26,51 @@ Other Enhancements
...
+code path. Particularly for functions with side effects, this was an undesired
+behavior and may have led to surprises.
+
+Now every group is evaluated only a single time.
+
+Previous behavior:
+
+.. code-block:: ipython
+
+    In [2]: df = pd.DataFrame({"a": ["x", "y"], "b": [1, 2]})
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: define df in a separate block above and show it

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_groupby.py (hunk 4 above)
Comment (fjetter) [reply pandas/tests/groupby/test_groupby.py]: I'll remove the links from this test since this test has been there before and is supposed to test whether the group name is available. The connection to this issue is merely coincidental.

For the (very similar) test in `groupby/test_apply.py` I added all these ticket numbers since you asked me to go through the issues and collect the tests. They are all duplicates and provide only very little additional information. All of them are linked to #2936 Is it ok if I remove all duplicates and only keep that one? same question for the whatsnew: list all duplicates or just the one reference issue?

**Comment by jreback** [comment]
i am ok with slightly duplicative tests if they are testing subtle behavior - if they are exact duplicates then ok to combine

**Comment by fjetter** [comment]
The `test_group_name_available_in_inference_pass` is now a subset of `test_group_apply_once_per_group` in `groupby/test_apply.py`. Although the implementation is very similar I'd keep them both since they test different behavior

File: doc/source/whatsnew/v0.25.0.rst [hunk 7]
```diff
@@ -26,6 +26,44 @@ Other Enhancements
...
+    In [6]: assert side_effects == ["x", "x", "y"]
+
+New behavior:
+
+.. ipython:: python
+
+    assert side_effects == ["x", "y"]
+
+
+(:issue:`2936`, :issue:`2656`, :issue:`7739`, :issue:`10519`, :issue:`12155`,
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: put the issue numbers at the end of the top paragraph.

File: doc/source/whatsnew/v0.25.0.rst [hunk 8]
```diff
@@ -26,6 +26,44 @@ Other Enhancements
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+GroupBy.apply on ``DataFrame`` evaluates first group only once
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The implementation of ``DataFrame.groupby.apply`` previously evaluated func
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: see if you can make this a proper reference, I think:

``:meth:`DataFrameGroupBy.apply() <pandas.core.groupby.DataFrameGroupBy.apply>` ``

should work

File: doc/source/whatsnew/v0.25.0.rst [hunk 9]
```diff
@@ -26,6 +26,44 @@ Other Enhancements
...
+    def func(group):
+        side_effects.append(group.name)
+        return group
+    df.groupby("a").apply(func)
+
+Previous behavior:
+
+.. code-block:: ipython
+
+    In [6]: assert side_effects == ["x", "x", "y"]
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: just print side_effects (and it will show the 3 values)

File: doc/source/whatsnew/v0.25.0.rst [hunk 10]
```diff
@@ -26,6 +26,44 @@ Other Enhancements
...
+
+New behavior:
+
+.. ipython:: python
+
+    assert side_effects == ["x", "y"]
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: same here, just show side_effects

File: pandas/_libs/reduction.pyx [hunk 11]
```diff
@@ -507,44 +507,41 @@ def apply_frame_axis0(object frame, object f, object names,
...
+            try:
+                if piece.index is chunk.index:
                     piece = piece.copy(deep='all')
                 else:
                     mutated = True
             except AttributeError:
                 pass
 
             results.append(piece)
+            if status > 0:
```
Comment (jreback) [inline pandas/_libs/reduction.pyx]: this notion of status is pretty opaque, can you add some comments and/or make this much more obvious what you are doing.

File: pandas/tests/groupby/test_apply.py [hunk 12]
```diff
@@ -101,10 +101,34 @@ def f(g):
     splitter = grouper._get_splitter(g._selected_obj, axis=g.axis)
     group_keys = grouper._get_group_keys()
```
Comment (jreback) [inline pandas/tests/groupby/test_apply.py]: big comment on what this is testing and why this is the correct result

File: pandas/tests/groupby/test_apply.py [hunk 13]
```diff
@@ -101,10 +101,34 @@ def f(g):
...
 
-    values, mutated = splitter.fast_apply(f, group_keys)
+    values, mutated, status = splitter.fast_apply(f, group_keys)
+    assert status == 0
     assert not mutated
 
 
+def test_group_apply_once_per_group():
+    # GH2936, GH7739, GH10519, GH2656, GH12155, GH20084, GH21417
```
Comment (jreback) [inline pandas/tests/groupby/test_apply.py]: add a 1-liner of what this is testing

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.25.0.rst [hunk 14]
```diff
@@ -26,6 +26,45 @@ Other Enhancements
...
 
+GroupBy.apply on ``DataFrame`` evaluates first group only once
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The implementation of :meth:`DataFrameGroupBy.apply() <pandas.core.groupby.DataFrameGroupBy.apply>`
+previously evaluated func consistently twice on the first group to infer if it
+is safe to use a fast code path. Particularly for functions with side effects,
+this was an undesired behavior and may have led to surprises.
+
+Now every group is evaluated only a single time.
```
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v0.25.0.rst]: it is affter the sentence above.

File: doc/source/whatsnew/v0.25.0.rst [hunk 15]
```diff
@@ -26,6 +26,45 @@ Other Enhancements
...
+    df = pd.DataFrame({"a": ["x", "y"], "b": [1, 2]})
+    df
+
+    side_effects = []
+    def func(group):
+        side_effects.append(group.name)
+        return group
+    df.groupby("a").apply(func)
+
+Previous behavior:
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: use ``*Previous Behaviour*`` (not the ``*``) and similar on New

File: pandas/_libs/reduction.pyx [hunk 16]
```diff
@@ -507,44 +507,41 @@ def apply_frame_axis0(object frame, object f, object names,
...
-            object.__setattr__(slider.dummy, 'name', names[i])
-            piece = f(slider.dummy)
-
-            # I'm paying the price for index-sharing, ugh
             try:
-                if piece.index is slider.dummy.index:
+                piece = f(chunk)
+            except:
+                raise InvalidApply('Let this error raise above us')
+            # Need to infer if low level index slider will cause segfaults
```
Comment (jreback) [inline pandas/_libs/reduction.pyx]: blank line here

File: pandas/_libs/reduction.pyx [hunk 17]
```diff
@@ -507,44 +507,41 @@ def apply_frame_axis0(object frame, object f, object names,
...
+            try:
+                if piece.index is chunk.index:
                     piece = piece.copy(deep='all')
                 else:
                     mutated = True
             except AttributeError:
                 pass
 
             results.append(piece)
+            if not successfull_fast_apply:
```
Comment (jreback) [inline pandas/_libs/reduction.pyx]: add a comment here on why you are returning early

File: pandas/core/groupby/ops.py [hunk 18]
```diff
@@ -165,26 +165,37 @@ def apply(self, f, data, axis=0):
         mutated = self.mutated
         splitter = self._get_splitter(data, axis=axis)
         group_keys = self._get_group_keys()
-
+        reuse_result = False
```
Comment (jreback) [inline pandas/core/groupby/ops.py]: this use of a flag is very odd and not very pythonic why is this needed at all?

File: pandas/core/groupby/ops.py [hunk 19]
```diff
@@ -165,26 +165,37 @@ def apply(self, f, data, axis=0):
...
             try:
-                values, mutated = splitter.fast_apply(f, group_keys)
-                return group_keys, values, mutated
+                result = splitter.fast_apply(f, group_keys)
+                fast_apply_result, mutated, successful_fast_apply = result
+                # If the fast apply path could be used we can return here.
+                # Otherwise we need to fall back to the slow implementation.
+                if successful_fast_apply:
+                    return group_keys, fast_apply_result, mutated
+                else:
```
Comment (jreback) [inline pandas/core/groupby/ops.py]: you don't need an else

File: pandas/core/groupby/ops.py (hunk 18 above)
Comment (jreback) [reply pandas/core/groupby/ops.py]: if you really really need this, then simply set ``result_values=None`` and test this

File: pandas/tests/groupby/test_apply.py [hunk 20]
```diff
@@ -101,10 +101,43 @@ def f(g):
...
+    # fall back to a slow apply code path for safety reasons.
+    assert succsessful_apply
     assert not mutated
 
 
+def test_group_apply_once_per_group():
+    # GH2936, GH7739, GH10519, GH2656, GH12155, GH20084, GH21417
+
+    # This test should ensure that a function is only evaluted
+    # once per group. Previously the function has been evaluated twice
```
Comment (jreback) [inline pandas/tests/groupby/test_apply.py]: this really covers *all* of the above issues? meaning they all have exactly the same input? I find that hard to believe. Please parameterize as needed, but run thru the cases.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_apply.py (hunk 20 above)
Comment (fjetter) [reply pandas/tests/groupby/test_apply.py]: The test cases are in fact pretty close to each others. To make the review easier I included all test cases but I believe most are redundant. If you see no need for the duplication, I can easily remove / shrink the test again

File: pandas/core/groupby/ops.py (hunk 18 above)
Comment (fjetter) [reply pandas/core/groupby/ops.py]: I've rewritten the logic to neither require the `status` identifier nor the `reuse_result` flag. The `reuse_result` flag was necessary in the previous implementation to catch the case for `MultiIndex` data which is handled by raising an exception in Cython. This could be handled differently if the multiindex check is performed outside of Cython but I didn't want to change too much about the current behavior. Having the index check close to the implementation seemed like a good idea to me

File: doc/source/whatsnew/v0.25.0.rst [hunk 21]
```diff
@@ -26,6 +26,44 @@ Other Enhancements
...
+    df = pd.DataFrame({"a": ["x", "y"], "b": [1, 2]})
+    df
+
+    side_effects = []
+    def func(group):
+        side_effects.append(group.name)
+        return group
+    df.groupby("a").apply(func)
+
+Previous Behaviour:
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you add a ``*Previous Behavior:*`` (and same for New below)

**Review by jreback — CHANGES_REQUESTED** [review]
very minor comment. pls merge master and ping on green.

File: pandas/core/groupby/ops.py [hunk 22]
```diff
@@ -165,26 +165,35 @@ def apply(self, f, data, axis=0):
         mutated = self.mutated
         splitter = self._get_splitter(data, axis=axis)
         group_keys = self._get_group_keys()
+        result_values = []
```
Comment (jreback) [inline pandas/core/groupby/ops.py]: I think better to make this: ``result_values=None``

File: pandas/core/groupby/ops.py [hunk 23]
```diff
@@ -165,26 +165,35 @@ def apply(self, f, data, axis=0):
...
                 # raise this error to the caller
                 pass
 
-        result_values = []
         for key, (i, group) in zip(group_keys, splitter):
             object.__setattr__(group, 'name', key)
 
+            # If the fast apply failed we may still use the result
+            # for the first group
+            if len(result_values) == 1 and i == 0:
```
Comment (jreback) [inline pandas/core/groupby/ops.py]: this would be then
``if result_values is None and i == 0``

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_apply.py [hunk 24]
```diff
@@ -102,9 +102,89 @@ def f(g):
...
 
     values, mutated = splitter.fast_apply(f, group_keys)
+
     assert not mutated
 
 
+@pytest.mark.parametrize(
+    "df,group_names",
+    [
+        (
```
Comment (WillAyd) [inline pandas/tests/groupby/test_apply.py]: Nitpicking but I find the readability here difficult. Can you start each set of parametrization with `(DataFrame(` and line break from there as required?

File: pandas/_libs/reduction.pyx [hunk 25]
```diff
@@ -507,40 +507,41 @@ def apply_frame_axis0(object frame, object f, object names,
...
                 else:
                     mutated = True
             except AttributeError:
                 pass
 
             results.append(piece)
+
+            # If the data was modified inplace we need to
+            # take the slow path to not risk segfaults
+            if require_slow_apply:
```
Comment (WillAyd) [inline pandas/_libs/reduction.pyx]: Might be overlooking but when would this ever be False?

File: doc/source/whatsnew/v0.25.0.rst [hunk 26]
```diff
@@ -63,6 +63,47 @@ is respected in indexing. (:issue:`24076`, :issue:`16785`)
...
     df['2019-01-01 12:00:00+04:00':'2019-01-01 13:00:00+04:00']
 
+.. _whatsnew_0250.api_breaking.groupby_apply_first_group_once:
+
+GroupBy.apply on ``DataFrame`` evaluates first group only once
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The implementation of :meth:`DataFrameGroupBy.apply() <pandas.core.groupby.DataFrameGroupBy.apply>`
+previously evaluated the supplied function consistently twice on the first group
+to infer if it is safe to use a fast code path. Particularly for functions with
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.25.0.rst]: Hmm I'm somewhat hesitant to market "side effects" as being fixed because I'm not sure we generally can make guarantees about mutations that occur during a groupby. Is it possible to leave this out and maybe use a relatively safe example like printing the groups instead?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.25.0.rst (hunk 26 above)
Comment (jreback) [reply doc/source/whatsnew/v0.25.0.rst]: I agree with @WillAyd  here, showing the printing is enough (and a bit more clear).

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
pls merge master

**Comment by jreback** [comment]
can you merge master and fix the conflict

**Comment by jreback** [comment]
can you merge master

**Comment by jreback** [comment]
rebased & whatsnew updated.

File: pandas/_libs/reduction.pyx (hunk 25 above)
Comment (jreback) [reply pandas/_libs/reduction.pyx]: fixed up

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@WillAyd if you have any more comments.

**Comment by WillAyd** [comment]
Thanks Jeff I’ll take a look tonight and merge if good

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
CI failure looked unrelated. Merged master will merge on green

**Comment by WillAyd** [comment]
Thanks @fjetter !

**Comment by CMCDragonkai** [comment]
Is this true only for groupby.apply? I tried test with just `apply` on a standard dataframe, and I found that the applied function is still being executed twice for the first row. This is quite problematic for me as my applied function is a side-effect.

**Comment by fjetter** [comment]
@CMCDragonkai This issue is only fixed for the groupby.apply path. The ordinary `DataFrame.apply` uses different code paths where maybe a similar issue is buried.

**Comment by CMCDragonkai** [comment]
Are you serious! So while groupby apply no longer has duplicate side effects, but normal apply still does!? Shouldn't it be easy to apply this fix to the situation for normal dataframe apply?

**Comment by fjetter** [comment]
@CMCDragonkai I'd need to check what is causing the duplicate execution to say for sure. My gut tells me that it should be similarly possible. I can not guarantee it without actually checking it.

**Comment by AlonMKovrr** [comment]
Are there any news regarding fixing this for the df.apply?

**Comment by alonme** [comment]
Are there any news regarding fixing this for the df.apply?

**Comment by MarcoGorelli** [comment]
> Are you serious! So while groupby apply no longer has duplicate side effects, but normal apply still does!?

@CMCDragonkai this is no way to address a community contributor who has volunteered their time to submit a pull request, please refrain from using such a tone in the future - thank you :)
